### PR TITLE
MGMT-7282: Operator upgrade CI job

### PR DESCRIPTION
**Pull request by pawanpinjarkar**
# Assisted Pull Request

## Description

For the operator upgrade CI job, we need to set some env vars during the initial part of the upgrade and later change the values after the upgrade. Setting the values in the scripts with this PR seems the easiest and most straightforward option for CI use. Also, it can be used for local testing.
 
This PR adds new variables to be used by the operator upgrade CI job.

- `ASSISTED_UPGRADE_OPERATOR` -  This is useful to determine which `ASSISTED_OPENSHIFT_VERSION` to be used. Before upgrading the operator, the value of this variable is `true`, which signals that the current operator needs to be updated. The value is derived from the default_ocp_versions.json. When set to true, `ASSISTED_OPENSHIFT_VERSION` gets the value `4.8`. Otherwise, `4.9`. After upgrading the operator, the value of this variable is `false`.

- `ASSISTED_STOP_AFTER_AGENT_DISCOVERY` - This is useful when waiting for a full cluster deployment is not needed. For example, during an upgrade CI job, we may not need to wait for the full deployment in some of the phases of the job. Before upgrading the operator, the value of this variable is `false`, which preserves the existing behavior.
After upgrading the operator, the value of this variable is `true`.

- `ASSISTED_CLUSTER_NAME`, `ASSISTED_CLUSTER_DEPLOYMENT_NAME`, `ASSISTED_AGENT_CLUSTER_INSTALL_NAME`, `ASSISTED_INFRAENV_NAME` - These are used for the upgraded cluster.

Consumed by: https://github.com/openshift/release/pull/21539

## List all the issues related to this PR

- [ ] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [X] CI/CD 

## What environments does this code impact?

- [X] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [X] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [] No tests needed

## Assignees

/cc @lranjbar 
/cc @flaper87 

## Checklist

- [X] Title and description added to both, commit and PR.
- [X] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [X] Reviewers have been listed
- [X] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: deploy/operator/after_upgrade.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+export UPGRADE_OPERATOR="false"
```
Comment (lranjbar) [inline deploy/operator/after_upgrade.sh:2]: I'm pretty bad at naming, but I'm wondering if this is too generic. Maybe `IS_OPERATOR_UPGRADE_FLOW` or `UPGRADE_ASSISTED_OPERATOR`? Since this is used to denote select which OCP version to start with.

Just trying to think about reading this. 🤔

File: deploy/operator/before_upgrade.sh [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+export SERVICE_IMAGE=""
+export INDEX_IMAGE="quay.io/ocpmetal/assisted-service-index:ocm-2.3"
```
Comment (lranjbar) [inline deploy/operator/before_upgrade.sh:3]: Where do we set the `INDEX_IMAGE` to be the newly built operator? I didn't see another value in `after_upgrade.sh`

**Review by lranjbar — COMMENTED** [review]
This looks mostly right to me, just a couple of questions.

File: deploy/operator/before_upgrade.sh (hunk 2 above)
Comment (pawanpinjarkar) [reply deploy/operator/before_upgrade.sh:3]: There are 3 phases in the upgrading job, before the upgrade, upgrade and after the upgrade. So, when `baremetalds-assisted-operator-upgrade-ref.yaml` calls  `baremetalds-assisted-operator-upgrade-commands.sh`, I am setting the latest index image and patching the catalogsource with that image.
https://github.com/openshift/release/pull/21539/files#diff-27e52a9f8e18f6f4a309185c25377a2839112270b55e25429bca81156182bd51R35

File: deploy/operator/after_upgrade.sh (hunk 1 above)
Comment (pawanpinjarkar) [reply deploy/operator/after_upgrade.sh]: `UPGRADE_ASSISTED_OPERATOR` sounds good.

File: deploy/operator/common.sh [hunk 3]
```diff
@@ -25,14 +25,21 @@ export HIVE_NAMESPACE="${HIVE_NAMESPACE:-hive}"
 # Versions #
 ############
 
+export UPGRADE_ASSISTED_OPERATOR="${UPGRADE_ASSISTED_OPERATOR:-false}"
 export OPENSHIFT_VERSIONS="${OPENSHIFT_VERSIONS:-$(cat ${__root}/data/default_ocp_versions.json)}"
 
 if [ -z "${ASSISTED_OPENSHIFT_INSTALL_RELEASE_IMAGE:-}" ]; then
-    export ASSISTED_OPENSHIFT_INSTALL_RELEASE_IMAGE=$(echo ${OPENSHIFT_VERSIONS} | jq -rc '[.[].release_image]|max')
+    if [ "${UPGRADE_ASSISTED_OPERATOR}" = "true" ]; then
+        export ASSISTED_OPENSHIFT_INSTALL_RELEASE_IMAGE=$(echo ${OPENSHIFT_VERSIONS} | jq -rc '[.[].release_image][-2]')
```
Comment (YuviGold) [inline deploy/operator/common.sh:33]: why `-2`? I don't understand these code section. In any case, we must not depend on indexing that might get change from time to time. (e.g. 4.6 and 4.7 will soon be removed)

File: deploy/operator/after_upgrade.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+export UPGRADE_ASSISTED_OPERATOR="false"
```
Comment (YuviGold) [inline deploy/operator/after_upgrade.sh:2]: Rename to be aligned with the same naming convention
```suggestion
export ASSISTED_UPGRADE_OPERATOR="false"
```

File: deploy/operator/after_upgrade.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+export ASSISTED_STOP_AFTER_AGENT_DISCOVERY="true"
```
Comment (YuviGold) [inline deploy/operator/after_upgrade.sh:3]: I think that it would be more intuitive if we will inverse it. Similar to what we have in assisted-test-infra. There's `deploy_nodes` and `deploy_nodes_with_install`
```suggestion
export ASSISTED_WAIT_FOR_INSTALLATION="false"
```
WDYT? @flaper87

File: deploy/operator/before_upgrade.sh [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (YuviGold) [inline deploy/operator/before_upgrade.sh:1]: The logic of the upgrade process shouldn't be split between repositories. I can see `before_upgrade.sh` and `after_upgrade.sh` here but not their usage.

Please move https://github.com/openshift/release/pull/21539/files#diff-27e52a9f8e18f6f4a309185c25377a2839112270b55e25429bca81156182bd51R20-R36 logic to be a `./deploy/operator/upgrade.sh` script maintained under the assisted-service repository. This way it would be accessible locally as well.

File: deploy/operator/before_upgrade.sh (hunk 2 above)
Comment (lranjbar) [reply deploy/operator/before_upgrade.sh:3]: Ah right it is more clear seeing it in code than talking through it. Okay so we will need to ability to override these exports therefore we will need to pass in the variables like we do in other places like:

`export INDEX_IMAGE=$INDEX_IMAGE-:quay.io/ocpmetal/assisted-service-index:ocm-2.3`

But for all of the exports to be safe

File: deploy/operator/common.sh (hunk 3 above)
Comment (lranjbar) [reply deploy/operator/common.sh:33]: Before the upgrade we are looking to install the version prior to the most current one of OCP. The most current version of OCP we are installing is 4.9, and the version previous to that is 4.8.

TLDR of the current flow is:
- Install hub cluster
- Install operator
- Install OCP 4.8 SNO spoke cluster
- Update operator
- Install OCP 4.9 SNO spoke cluster

Comment (pawanpinjarkar) [reply deploy/operator/common.sh:33]: Thank you, Lisa. I can add comments explaining the logic but I don't see any other alternative atm.

File: deploy/operator/after_upgrade.sh (hunk 4 above)
Comment (pawanpinjarkar) [reply deploy/operator/after_upgrade.sh]: Done

File: deploy/operator/common.sh (hunk 3 above)
Comment (pawanpinjarkar) [reply deploy/operator/common.sh:42]: Added comments in the script explaining the need and logic.

File: deploy/operator/before_upgrade.sh (hunk 6 above)
Comment (pawanpinjarkar) [reply deploy/operator/before_upgrade.sh:1]: Done. Moved to `./deploy/operator/upgrade.sh`. `before_upgrade.sh` and `after_upgrade.sh` are to be used by CI job. My priority is to get the job ready before feature freeze and then focus on running the upgrade job locally.

File: deploy/operator/common.sh (hunk 3 above)
Comment (YuviGold) [reply deploy/operator/common.sh:42]: Thank you for the clarification

File: deploy/operator/before_upgrade.sh (hunk 6 above)
Comment (YuviGold) [reply deploy/operator/before_upgrade.sh:1]: Great.
Now looking at the CI job, I don't understand why do we need 3 different steps. I'd appreciate if also `after_upgrade` step could be as part of the `upgrade` script.
https://github.com/openshift/release/pull/21539/files#diff-10db92a2ad655eae770c3c0383bf65fa41baf1fe82a8a388e00ea02634e02199R53-R57

The same goes for ZTP. We verify our installation stages as part of the ZTP script.
Then, all the variables will be assigned inside the script and the `after_upgrade.sh` and `before_upgrade.sh` won't be necessairy. In case someone would want to override them - then it should be possible.

Comment (pawanpinjarkar) [reply deploy/operator/before_upgrade.sh:1]: I suggest making these code style changes later and get the CI job-ready first. I will create a follow-up PR later.

**Comment by lranjbar** [comment]
There has been a decent amount of churn on this PR and it is in a working state so I'm ready to merge it to make it easier to test in `openshift/release`

The following need to be addressed in a follow up PRs:
- How are we piping in the openshift releases into `ASSISTED_OPENSHIFT_INSTALL_RELEASE_IMAGE`? I think this stanza needs to be tested: https://github.com/openshift/release/pull/21539/files#diff-b4ec9980e779ed206c8f8c154f750816537b6dde361289606477872c7fd319aaR27-R28 I'm not entirely convinced we wont need `ASSISTED_OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE` but also I believe it will be easier to test this after this PR merges.
- Reusing the same worker node vs having two worker nodes this can be moved to another JIRA issue as an enhancement to the upgrade job.
